### ISSUE TITLE
NWNX_EVENTS - TRAP

### ISSUE THREAD

**Issue by Christian-Vassallo** [nwnxee/unified#1462]
Hello guys,

since what we said on Discord and, apparently, a well known issue, I would like to report the problem of Trap Events.
Indeed they are fired theee times: when I ask to do the action, during the action, at the end of the action.

the action result flag is always 2 during a normal operation (I.e not stopping or failing) and it’s not possible to use this script natively without any other workaround.

Some idea to fix it in order to fire it correctly once?

**Comment by Daztek** [comment]
It's just how those hooked functions work, TrapDisarm for example, when it runs the first time, it'll check if the trap is within use distance, if it isn't, it'll queue a moveto action and the disarm trap action again. It'll do the moveto action and then it'll run the DisarmTrap action again where it'll queue the animation and another DisarmTrap action, when the animation action is done it'll run DisarmTrap a final time and do the actual disarming.

The actual trap stuff happens after the animation stuff has played so you could possible check if `CNWSCreature::m_bTrapAnimationPlayed` is true before firing the event, but I haven't tried.

**Comment by Daztek** [comment]
Or better, add push a new `TRAP_ANIMATION_PLAYED` eventdata so the user can decide what to do

**Comment by Christian-Vassallo** [comment]
Not really useful because there was no way to get animation data, without writing new code on the plug-in itself. 
Your solution is fine in case you want to re-write the action system which is not my case (and I guess the one in general).
Another point is that it’s difficult to determine if the distance is not enough to interact with the trap in every case. In fact, the trap is rectangular and the distance to the center might vary and it’s not the same than the distance to the shape limits. So you might have distances where you require to move which are greater than distances where you don’t need to move (example at the corner of the rectangular).
Clearly everything might be fixed by using a threshold which is a dirty solution in my opinion.
The last point is that I am not sure how RESULT variable should work because I made several test (critical failure, stopping, interrupting) and in every case the result was 2.
I also tried to analyze the animation of the trap by NWNX OBJECT but the result was always equal to 0.

Finally a workaround it’s possible by using the timing bar events (start and stop) but, again, it’s a dirty choice.

I didn’t check how the code of the plug-in is written yet, however I guess that there is something buggy
